### PR TITLE
Prevent cross origin upvoting

### PR DESCRIPTION
**Pull request by pietvanzoen**
... plus other fixes.

### CODE REVIEW

File: .eslintrc.json [hunk 1]
```diff
@@ -1,10 +1,7 @@
 {
-  "extends": [
-    "eslint:recommended",
-    "prettier"
-  ],
+  "extends": ["eslint:recommended", "prettier"],
   "parserOptions": {
-    "ecmaVersion": 8
+    "ecmaVersion": 2018
```
Comment (keepitterron) [inline .eslintrc.json:4]: watchout.badass.meme

File: server/middleware.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+function assertOriginMatchesLocation(req, res, next) {
+  const origin = url.parse(req.header("origin") || "");
+  const location = url.parse(req.body.location);
+  if (origin.host !== location.host) {
+    log.debug(
+      `Origin '${origin.host}' did not match host for URL '${location.href}'`
+    );
+    return next(
+      new errs.ForbiddenError("Cross origin upvoting is not allowed.")
```
Comment (keepitterron) [inline server/middleware.js:13]: 🎉

File: server/middleware.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  if (req.body && req.body.location) {
+    req.body.location = decodeURIComponent(req.body.location);
+  }
+  if (req.params && req.params.location) {
+    req.params.location = decodeURIComponent(req.params.location);
+  }
+  next();
+}
+
+function catchAsyncErrors(callback) {
```
Comment (keepitterron) [inline server/middleware.js:29]: niiiiiice

File: server/routes.js [hunk 4]
```diff
@@ -3,42 +3,55 @@ const restify = require("restify");
...
-        .then(resp => {
-          res.send(resp);
-          next();
-        })
-        .catch(err => {
-          res.send(err);
-          next();
-        });
-    }
+    restify.plugins.throttle({
```
Comment (keepitterron) [inline server/routes.js:43]: what does the throttle do? too lazy to look it up :/

**Review by keepitterron — APPROVED** [review]
✨ ✨ ✨ ✨ 💖 👍 WUNDERBAR! 👍 💖 ✨ ✨ ✨ ✨

File: .eslintrc.json (hunk 1 above)
Comment (pietvanzoen) [reply .eslintrc.json:4]: 😎

File: server/routes.js (hunk 4 above)
Comment (pietvanzoen) [reply server/routes.js:43]: prevents some spamming. default limits requests per ip address to 1 per 2 seconds. you can adjust the rate with the env variable.

**Pull request merged**